lib/datasource/crate/index.ts

```diff
@@ -1,43 +1,168 @@
+  info: RegistryInfo,
+  lookupName: string
+): Promise<string> {
+  if (info.clonePath) {
+    const path = join(info.clonePath, ...getIndexSuffix(lookupName));
+    return fs.readFile(path, { encoding: 'utf8' });
+  }
+
+  if (info.flavor === RegistryFlavor.CratesIo) {
+    const crateUrl = CRATES_IO_BASE_URL + getIndexSuffix(lookupName).join('/');
+    try {
+      return (await http.get(crateUrl)).body;
+    } catch (err) {
+      if (
+        err.statusCode === 429 ||
+        (err.statusCode >= 500 && err.statusCode < 600)
+      ) {
+        throw new ExternalHostError(err);
+      }
+    }
+  }
+
+  throw new Error(`unsupported crate registry flavor: ${info.flavor}`);
+}
+
+/// Computes the dependency URL for a crate, given
+/// registry information
+function getDependencyUrl(info: RegistryInfo, lookupName: string): string {
+  switch (info.flavor) {
+    case RegistryFlavor.CratesIo:
+      return `https://crates.io/crates/${lookupName}`;
+    case RegistryFlavor.Cloudsmith: {
+      // input: https://dl.cloudsmith.io/basic/$org/$repo/cargo/index.git
+      const tokens = info.url.pathname.split('/');
+      const org = tokens[2];
+      const repo = tokens[3];
+      return `https://cloudsmith.io/~${org}/repos/${repo}/packages/detail/cargo/${lookupName}`;
+    }
+    default:
+      return `${info.rawUrl}/${lookupName}`;
+  }
+}
+
+/// Fetches information about a registry, by url.
+/// If no url is given, assumes crates.io.
+/// If an url is given, assumes it's a valid Git repository
+/// url and clones it to cache.
+async function fetchRegistryInfo(
+  registryUrl: string
+): Promise<RegistryInfo | null> {
+  let url: URL;
+  try {
+    url = new URL(registryUrl);
+  } catch (err) {
+    logger.debug({ registryUrl }, 'could not parse registry URL');
+    return null;
+  }
+
+  let flavor: RegistryFlavor;
+  if (url.hostname === 'crates.io') {
+    flavor = RegistryFlavor.CratesIo;
+  } else if (url.hostname === 'dl.cloudsmith.io') {
+    flavor = RegistryFlavor.Cloudsmith;
+  } else {
+    flavor = RegistryFlavor.Other;
+  }
+
+  const registry: RegistryInfo = {
+    flavor,
+    rawUrl: registryUrl,
+    url,
+  };
+
+  if (flavor !== RegistryFlavor.CratesIo) {
+    let clonePath = registryClonePaths[registryUrl];
+    if (!clonePath) {
+      clonePath = await ensureCacheDir(`crate-registry-${url.hostname}`);
+      logger.info({ clonePath, registryUrl }, `Cloning private cargo registry`);
+      {
+        const git = Git();
+        await git.clone(registryUrl, clonePath, {
+          '--depth': 1,
+        });
+      }
+      registryClonePaths[registryUrl] = clonePath;
+    }
+    registry.clonePath = clonePath;
+  }
+
+  return registry;
+}
+
 export async function getReleases({
   lookupName,
+  registryUrl,
 }: GetReleasesConfig): Promise<ReleaseResult | null> {
   const cacheNamespace = 'datasource-crate';
-  const cacheKey = lookupName;
+  const cacheKey = registryUrl ? `${registryUrl}/${lookupName}` : lookupName;
```

Comment:
This could be simplified - `registryUrl` should never be undefined anymore here.

In the case of `crates.io` crates, it's just `https://crates.io`

Reply:
Simplified in https://github.com/renovatebot/renovate/pull/8083/commits/ba45a2d88c56460be17be549692d45886a3c900e